### PR TITLE
Getproducts

### PR DESCRIPTION
**Pull request by tavisca-sgogawale**

### CODE REVIEW

File: src/OnlineRetailPortal.Core/Models/GetProductCoreResponce.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace OnlineRetailPortal.Core
+{
+    public class GetProductsCoreResponce
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/GetProductCoreResponce.cs]: We do not need response in core model,
We can create getproductresponse at service level using core models 
You don’t need explicit core response class

File: src/OnlineRetailPortal.Core/Models/Price.cs [hunk 2]
```diff
@@ -8,6 +8,6 @@ public class Price
     {
         public double Amount { get; set; }
         public bool IsNegotiable { get; set; }
-        public Currency Currency { get; }
+        public string Currency { get; set; }
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Price.cs]: Add a note that this is being handled with other pr .. 
aditi is working on these changes for bundling currency and amount

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 3]
```diff
@@ -38,7 +38,7 @@ public Product(IProductStore productStore)
             this.productStore = productStore;
         }
 
-        public async Task<GetProductsServiceResponse> GetProductsAsync(GetProductsServiceRequest serviceRequest)
+        public static async Task<GetProductsCoreResponce> GetProductsAsync(GetProductsServiceRequest serviceRequest)
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs]: This should be getproductsresponse 
You need not to add a word core with it as this class is not tied up with core models
Also why the method is static

File: src/OnlineRetailPortal.Mock/Models/Address.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Mock
+{ 
+    public class Address
+    {
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Mock/Models/Address.cs]: Why do we need this in mock

File: src/OnlineRetailPortal.Core/Translators/GetProductServiceResponseTranslator.cs [hunk 5]
```diff
@@ -8,35 +8,34 @@ namespace OnlineRetailPortal.Core
...
-                    HeroImage = new Contracts.Image() { Url = getProductResponse.Product.HeroImage.Url },
+                    HeroImage = new Image() { Url = getProductResponse.Product.HeroImage.Url },
                     ExpirationDate = getProductResponse.Product.ExpirationDate,
                     PostDateTime = getProductResponse.Product.PostDateTime,
                     Description = getProductResponse.Product.Description,
-                    Price = new Contracts.Price() { Amount = getProductResponse.Product.Price.Amount, IsNegotiable = getProductResponse.Product.Price.IsNegotiable , Currency = getProductResponse.Product.Price.Currency},
+                    Price = new Price() { Amount = getProductResponse.Product.Price.Amount, IsNegotiable = getProductResponse.Product.Price.IsNegotiable , Currency = getProductResponse.Product.Price.Currency},
                     PurchasedDate = getProductResponse.Product.PurchasedDate,
-                    PickupAddress = new Contracts.Address()
+                    PickupAddress = new Address()
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Translators/GetProductServiceResponseTranslator.cs]: Add null checks on optional fields

File: src/OnlineRetailPortal.Mock/Models/Price.cs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿namespace OnlineRetailPortal.Mock
+{
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Mock/Models/Price.cs]: Remove all the duplicate models declared in mock project.. 
why do we need them?

File: src/OnlineRetailPortal.Mock/Models/MockProductDatabase.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace OnlineRetailPortal.Mock.Models
+{
+
+    public class MockProductDatabase
+    {
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Mock/Models/MockProductDatabase.cs]: Where are we using this class?

**Review by Tavisca-smohite — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Core/Models/GetProductCoreResponce.cs (hunk 1 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/GetProductCoreResponce.cs]: Yes but we are giving paging info in response so GetProductCoreResponse consist of List<Product> PagingInfo properties

File: src/OnlineRetailPortal.Core/Models/Price.cs (hunk 2 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Price.cs]: You mean to say there should be type Money which contains Amount and Currency properties ??

File: src/OnlineRetailPortal.Mock/Models/Price.cs (hunk 6 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Mock/Models/Price.cs]: I have created these models bcoz we are creating a List of Products in mock.

File: src/OnlineRetailPortal.Core/Models/Price.cs (hunk 2 above)
Comment (tavisca-swapnil) [reply src/OnlineRetailPortal.Core/Models/Price.cs]: Add a comment stating 'Amount' and 'Currency' will be later part of a single property. @Tavisca-smohite

File: src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs [hunk 8]
```diff
@@ -8,35 +8,40 @@ namespace OnlineRetailPortal.Core
...
-                    Images = getProductResponse.Product.Images.Select(x => new Contracts.Image
+                    Images = getProductResponse.Product.Images.Select(x => new Image
                     {
                         Url = x.Url
                     }).ToList(),
-                    Status = (Contracts.Status)getProductResponse.Product.Status.GetHashCode(),
-                    Category = (Contracts.Category)getProductResponse.Product.Category.GetHashCode()
-                }
+                    Status = (Status)getProductResponse.Product.Status,
+                    Category = (Category)getProductResponse.Product.Category
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs]: Casting should not be done in such a manner, try .ToStatusEntity/.ToStatusModel for translation.

The translation logic will be present in these (.ToStatusEntity/.ToStatusModel) methods

Comment (tavisca-swapnil) [reply src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs]: Same comment applies throughout

File: src/OnlineRetailPortal.Services/Services/ProductService.cs [hunk 9]
```diff
@@ -1,22 +1,25 @@
...
-    class ProductService : IProductService
+   public class ProductService : IProductService
     {
-       
-        public Task<GetProductsServiceResponse> GetProductsAsync(GetProductsServiceRequest request)
+        public async Task<GetProductsServiceResponse> GetProductsAsync(GetProductsServiceRequest request)
         {
-            throw new NotImplementedException();
+            IProductStore productStore=null;
+            var response = await Core.Product.GetProductsAsync(request,productStore );
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Services/Services/ProductService.cs]: Why are we sending null productstore?

File: src/OnlineRetailPortal.Services/Services/ProductService.cs [hunk 10]
```diff
@@ -1,22 +1,25 @@
...
+            //get list based on paging info 
+            return await Task.FromResult(response.ToGetProductsContract());
         }
 
-        public Task<GetProductServiceResponse> GetProductAsync(string productId)
+        public async Task<GetProductServiceResponse>  GetProductAsync(string productId)
         {
-            throw new NotImplementedException();
+            IProductStore productStore = null;
+           var response = await Core.Product.GetProductAsync(productId, productStore);
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Services/Services/ProductService.cs]: Why are we sending null productstore?

File: src/OnlineRetailPortal.Web/Models/Category.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        Properties,
+        Cars,
+        Furniture,
+        Mobiles,
+        Bikes,
+        Books,
+        Fashions,
+        Electronics,
+        Others
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Models/Category.cs]: weren't we supposed to keep them singular?

File: src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+                        Line1 = getProductServiceResponse.Product.PickupAddress.Line1,
+                        Line2 = getProductServiceResponse.Product.PickupAddress.Line2,
+                        Pincode = getProductServiceResponse.Product.PickupAddress.Pincode
+                    },
+                    Images = getProductServiceResponse.Product.Images.Select(x => new Web.Image
+                    {
+                        Url = x.Url
+                    }).ToList(),
+                    Status = (Web.Status)getProductServiceResponse.Product.Status.GetHashCode(),
+                    Category = (Web.Category)getProductServiceResponse.Product.Category.GetHashCode()
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs]: Are we hashing type X to get an int and casting it back to type X? If yes, do we need this?

**Review by tavisca-swapnil — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs (hunk 12 above)
Comment (Tavisca-smohite) [reply src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs]: Hashing Seems like over engineering to me ..
Shubham please comment the reason and come up with the simplest approach

File: src/OnlineRetailPortal.Services/Services/ProductService.cs (hunk 10 above)
Comment (Tavisca-smohite) [reply src/OnlineRetailPortal.Services/Services/ProductService.cs]: Shubham you need to add a comment that factory or logic to resolve the product store will be handled later ( or whatever applies here)

File: src/OnlineRetailPortal.Web/Models/Category.cs (hunk 11 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Web/Models/Category.cs]: Category was in future scope and it it already implemented by rishabh

File: src/OnlineRetailPortal.Services/Services/ProductService.cs (hunk 10 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Services/Services/ProductService.cs]: done

File: src/OnlineRetailPortal.Web/Models/Category.cs (hunk 11 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Web/Models/Category.cs]: done

File: src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs (hunk 12 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs]: done

File: src/OnlineRetailPortal.Mock/Models/MockProductDatabase.cs (hunk 7 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Mock/Models/MockProductDatabase.cs]: For mock implementation

File: src/OnlineRetailPortal.Services/Services/ProductService.cs (hunk 9 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Services/Services/ProductService.cs]: comment added it is in future scope

File: src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs (hunk 8 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs]: done

File: src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+                    PostDateTime = getProductResponse.Product.PostDateTime,
+                    Description = getProductResponse.Product.Description,          
+                    PurchasedDate = getProductResponse.Product.PurchasedDate,
+                    PickupAddress = new Address()
+                    {
+                        City = getProductResponse.Product.PickupAddress.City,
+                        State = getProductResponse.Product.PickupAddress.State,
+                        Line1 = getProductResponse.Product.PickupAddress.Line1,
+                        Line2 = getProductResponse.Product.PickupAddress.Line2,
+                        Pincode = getProductResponse.Product.PickupAddress.Pincode
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs]: Use .ToEntity()/.ToModel() for such translations

File: src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+            switch (status)
+            {
+                case Contracts.Status.Active:
+                    return Status.Active;
+                case Contracts.Status.Disabled:
+                    return Status.Disabled;
+                case Contracts.Status.Sold:
+                    return Status.Sold;
+                default:
+                    throw new NotSupportedException("This status is not available");
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs]: Can we change it to "This Status is not supported :" + status .

File: src/OnlineRetailPortal.Web/Controllers/ProductsController.cs [hunk 15]
```diff
@@ -19,15 +19,25 @@ public ProductsController(IProductService productService = null)
...
-        //   // response = await _productService.GetProductAsync(productId);
-        //    return response.ToDataContract();
-        //    //return new GetProductResponse();
-        //}
-
+        [HttpGet("products")]
+        public async Task<GetProductsResponse> GetProductsAsync(int pageNo, int pageSize)
+        {
+            var request = GetProductsServiceRequestTranslator.ToServiceRequest(pageNo, pageSize);
+            GetProductsRequestValidator validationRules = new GetProductsRequestValidator();
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Controllers/ProductsController.cs]: Rename validationRules to Validator

**Review by tavisca-swapnil — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Web/Controllers/ProductsController.cs (hunk 15 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Web/Controllers/ProductsController.cs]: done

File: src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs (hunk 14 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs]: done

File: src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs (hunk 13 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Translators/GetProductCoreResponseTranslator.cs]: done

File: src/OnlineRetailPortal.Mock/MockProductStore.cs [hunk 16]
```diff
@@ -69,8 +68,18 @@ public async Task<GetProductsStoreResponse> GetProductsAsync(GetProductsStoreReq
 
             response = await Task.Run(() =>
             {
-                List<Product> products = productList.Where(y => y.Status == Status.Active).OrderBy(x => x.PostDateTime).Take(50).ToList();
-                response = products.ToGetProductsStoreResponse();
+                List<Product> products = productList.OrderByDescending(x => x.PostDateTime).ToList();
+                int startIndex = (request.PagingInfo.PageNumber - 1) * (request.PagingInfo.PageSize);
+                int endIndex = startIndex + request.PagingInfo.PageSize - 1;
+                List<Product> newProductList = new List<Product>();
+                for (int i = startIndex; i <= endIndex; i++)
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Mock/MockProductStore.cs]: Replace 'i' with 'currentIndex'

File: src/OnlineRetailPortal.Mock/MockProductStore.cs [hunk 17]
```diff
@@ -69,8 +68,18 @@ public async Task<GetProductsStoreResponse> GetProductsAsync(GetProductsStoreReq
 
             response = await Task.Run(() =>
             {
-                List<Product> products = productList.Where(y => y.Status == Status.Active).OrderBy(x => x.PostDateTime).Take(50).ToList();
-                response = products.ToGetProductsStoreResponse();
+                List<Product> products = productList.OrderByDescending(x => x.PostDateTime).ToList();
+                int startIndex = (request.PagingInfo.PageNumber - 1) * (request.PagingInfo.PageSize);
+                int endIndex = startIndex + request.PagingInfo.PageSize - 1;
+                List<Product> newProductList = new List<Product>();
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Mock/MockProductStore.cs]: Rename 'newProductIndex'.

File: src/OnlineRetailPortal.Mock/MockProductStore.cs [hunk 18]
```diff
@@ -69,8 +68,18 @@ public async Task<GetProductsStoreResponse> GetProductsAsync(GetProductsStoreReq
...
+                for (int i = startIndex; i <= endIndex; i++)
+                {
+                    if (i == products.Count())
+                        break;
+                    newProductList.Add(products[i]);
+                }
+                request.PagingInfo.TotalPages = productList.Count() / request.PagingInfo.PageSize+ productList.Count()% request.PagingInfo.PageSize;
+                response = newProductList.ToGetProductsStoreResponse(request.PagingInfo.PageNumber,request.PagingInfo.PageSize, request.PagingInfo.TotalPages);
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Mock/MockProductStore.cs]: Why are we not sending request.PagingInfo and sending every field inside it seperately instead.

**Review by tavisca-swapnil — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Mock/MockProductStore.cs (hunk 18 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Mock/MockProductStore.cs]: done

File: src/OnlineRetailPortal.Mock/MockProductStore.cs (hunk 17 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Mock/MockProductStore.cs]: done

File: src/OnlineRetailPortal.Mock/MockProductStore.cs (hunk 16 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Mock/MockProductStore.cs]: done

**Review by tavisca-swapnil — APPROVED** [review]

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 19]
```diff
@@ -3,30 +3,48 @@
...
+        private static IProductStore _productStore;
 
-        public Product(string sellerId, string name, Price price)
+        public Product(Price price, string sellerId ,string name)
+        {
+            Price = price;
+            SellerId = sellerId;
+            Name = name;
+        }
+        public static async Task<ProductsWithPageInitiation>  GetProductsAsync(GetProductsServiceRequest serviceRequest, IProductStore productStore)
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs:31]: what this method do ?
and why there is static _productstore property in core Product model

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 20]
```diff
@@ -3,30 +3,48 @@
...
+        {
+            _productStore = productStore;
+            GetProductsStoreRequest request = serviceRequest.ToProductStoreRequest();
+            GetProductsStoreResponse response = await _productStore.GetProductsAsync(request);
+            return response.ToProductsWithPageInitiation();
+        }
+       
+        public static async Task<Product> GetProductAsync(string productId, IProductStore productStore)
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs]: Method name should be `GetAsync`

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 21]
```diff
@@ -3,30 +3,48 @@
...
         {
-            this.SellerId = sellerId;
-            this.Name = name;
-            this.Price = price;
+            _productStore = productStore;
+            GetProductStoreResponse response = await _productStore.GetProductAsync(productId);
+            return response.ToGetProductServiceResponse();
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs]: this should be `response.ToProduct()`

File: src/OnlineRetailPortal.Core/Models/ProductsWithPageInitiation.cs [hunk 22]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace OnlineRetailPortal.Core
+{
+    public class ProductsWithPageInitiation
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/ProductsWithPageInitiation.cs:5]: this is not needed in the core model

File: src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+            return new Product().PickupAddress = new Address()
+            {
+                Line1 = pickupAddress.Line1,
+                Line2 = pickupAddress.Line2,
+                City = pickupAddress.City,
+                Pincode = pickupAddress.Pincode,
+                State = pickupAddress.State
+            };
+        }
+        public static DateTime? ToEntity(this DateTime? purchasedDate)
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs]: remove this method

File: src/OnlineRetailPortal.Web/Validations/GetProductsRequestValidator.cs [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace OnlineRetailPortal.Web
+{
+    public class GetProductsRequestValidator : AbstractValidator<GetProductsServiceRequest>
+    {
+        public GetProductsRequestValidator()
+        {
+            RuleFor(x => x.PagingInfo.PageNumber)
+            .Cascade(CascadeMode.StopOnFirstFailure)
+            .NotNull()
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Web/Validations/GetProductsRequestValidator.cs:18]: pagenumber and pagesize were int in my opinion .Also as per contract , these fields are optional . Why we need validation on these fields then ?

**Review by Tavisca-smohite — CHANGES_REQUESTED** [review]

**Review by achirametavisca — APPROVED** [review]

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 19 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs:31]: done

File: src/OnlineRetailPortal.Web/Validations/GetProductsRequestValidator.cs (hunk 24 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Web/Validations/GetProductsRequestValidator.cs:18]: done

File: src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs (hunk 23 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Web/Translators/GetProductResponseTranslator.cs]: done

File: src/OnlineRetailPortal.Core/Models/ProductsWithPageInitiation.cs (hunk 22 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/ProductsWithPageInitiation.cs:5]: as discussed we are keeping it.

**Comment by achirametavisca** [comment]
You can merge this PR

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 21 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs]: done

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 20 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs]: done

**Pull request merged**